File: 03-backtracking-generation/binary-backtracking-2.cpp
#include<stdio.h>

int n, count = 1;

void init(int b[]) {
    int j;
    for (j = 1; j <= n; j++) b[j] = 1;
}

void out(int s[]) {
    printf("%d:  ", count);
    int i;
    for (i = 1; i <= n; i++) printf("%d  ", s[i]);
    count++;
    printf("\n");
}

void Try(int i, int b[], int s[]) {
    int j;
    for (j = 1; j <= n; j++) {
        if (b[j] == 1) {
            s[i] = j;
            b[j] = 0;
            if (i == n) out(s);
            else Try (i + 1, b, s);
            b[j] = 1;
        }
    }
}

int main() {
    printf("input n: ");
    scanf("%d", &n);
    int b[n], s[n];
    init(b);
    Try(1, b, s);
}

File: 03-backtracking-generation/binary-backtracking-2.py
def init(n):
    return [1] * n

def out(s):
    global count
    print(f"{count}:  ", end="")
    for i in range(n):
        print(f"{s[i]}  ", end="")
    count += 1
    print()

def try_permutations(i, b, s):
    for j in range(n):
        if b[j] == 1:
            s[i] = j + 1
            b[j] = 0
            if i == n - 1:
                out(s)
            else:
                try_permutations(i + 1, b, s)
            b[j] = 1

def main():
    global n
    global count
    count = 1
    n = int(input("input n: "))
    b = init(n)
    s = [0] * n
    try_permutations(0, b, s)

if __name__ == "__main__":
    main()
